src/main/java/com/aws/greengrass/shadowmanager/sync/model/MergedFullShadowSyncRequest.java

```diff
@@ -0,0 +1,143 @@
+     */
+    public MergedFullShadowSyncRequest(String thingName, String shadowName,
+                                       RequestMerger merger, SyncRequest... mergedRequests) {
+        super(thingName, shadowName);
+        this.merger = Objects.requireNonNull(merger);
+        if (mergedRequests != null && mergedRequests.length > 0) {
+            this.mergedRequests.addAll(flatten(Arrays.asList(mergedRequests)));
+        }
+    }
+
+    @Override
+    public void execute(SyncContext context)
+            throws RetryableException, SkipSyncRequestException, InterruptedException, UnknownShadowException {
+        super.setContext(context);
+
+        if (getMergedRequests().stream().anyMatch(r -> r.getClass().equals(FullShadowSyncRequest.class))) {
+            // If a full sync has been directly requested, not due to merging of requests, respect it.
+            super.execute(context);
+            return;
+        }
+
+        List<SyncRequest> necessaryMergedUpdates = getNecessaryMergedRequests(context);
+        if (necessaryMergedUpdates.isEmpty()) {
+            logger.atDebug()
+                    .kv(LOG_THING_NAME_KEY, getThingName())
+                    .kv(LOG_SHADOW_NAME_KEY, getShadowName())
+                    .log("Ignoring unnecessary sync request");
+            return;
+        }
+
+        if (necessaryMergedUpdates.stream().allMatch(r -> r instanceof CloudUpdateSyncRequest)
+                || necessaryMergedUpdates.stream().allMatch(r -> r instanceof LocalUpdateSyncRequest)) {
+            SyncRequest consolidatedUpdateRequest = necessaryMergedUpdates.stream().reduce(merger::merge).get();
+            logger.atDebug()
+                    .kv(LOG_THING_NAME_KEY, getThingName())
+                    .kv(LOG_SHADOW_NAME_KEY, getShadowName())
+                    .log("Full sync not needed, executing a "
+                            + consolidatedUpdateRequest.getClass().getSimpleName() + " instead");
+            consolidatedUpdateRequest.execute(context);
+            return;
+        }
+
+        super.execute(context);
+    }
+
+    /**
+     * Create a list of all the merged requests that require execution,
+     * as deemed by {@link BaseSyncRequest#isUpdateNecessary(SyncContext)}.
+     *
+     * @param context sync context
+     * @return sync requests
+     * @throws RetryableException       When error occurs in sync operation indicating a request needs to be retried
+     * @throws SkipSyncRequestException When error occurs in sync operation indicating a request needs to be skipped.
+     * @throws UnknownShadowException   When shadow not found in the sync table.
+     */
+    private List<SyncRequest> getNecessaryMergedRequests(SyncContext context)
+            throws RetryableException, UnknownShadowException, SkipSyncRequestException {
+        List<SyncRequest> necessaryUpdates = new ArrayList<>();
+        for (SyncRequest request : mergedRequests) {
```

Comment:
```suggestion
        return mergedRequests.stream()
          .filter(r -> request instanceof BaseSyncRequest)
          .filter(r -> ((BaseSyncRequest) request).isUpdateNecessary(context))
          .collect(Collectors.toList()); // Or your can readily return the stream
```

Reply:
rob had this suggestion too, isUpdateNecessary throws and we dont want to wrap it in a runtime exception